test/framework/control_plane.go

```diff
@@ -34,84 +34,191 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-const (
-	// eventuallyInterval is the polling interval used by gomega's Eventually
-	eventuallyInterval = 10 * time.Second
+// Package globals used to populate default values.
+// Can be updated by consumer to change default values.
+var (
+	// DefaultTimeout is used for the default timeout for every `Eventually` call.
+	DefaultTimeout = 10 * time.Minute
+
+	// DefaultPollInterval is used as the default for how frequently `Eventually` polls.
+	DefaultPollInterval = 10 * time.Second
 )
 
-// ControlplaneClusterInput defines the necessary dependencies to run a multi-node control plane cluster.
-type ControlplaneClusterInput struct {
-	Management        ManagementCluster
-	Cluster           *clusterv1.Cluster
-	InfraCluster      runtime.Object
-	Nodes             []Node
-	MachineDeployment MachineDeployment
-	RelatedResources  []runtime.Object
-	CreateTimeout     time.Duration
-	DeleteTimeout     time.Duration
+// Interfaces to scope down client.Client
 
-	ControlPlane    *controlplanev1.KubeadmControlPlane
-	MachineTemplate runtime.Object
+// Getter can get resources.
+type Getter interface {
+	Get(ctx context.Context, key client.ObjectKey, obj runtime.Object) error
+}
+
+// Creator can creates resources.
+type Creator interface {
+	Create(ctx context.Context, obj runtime.Object, opts ...client.CreateOption) error
+}
+
+// Lister can lists resources.
+type Lister interface {
+	List(ctx context.Context, list runtime.Object, opts ...client.ListOption) error
+}
+
+// Deleter can delete resources.
+type Deleter interface {
+	Delete(ctx context.Context, obj runtime.Object, opts ...client.DeleteOption) error
+}
+
+// ListDeleter combines interfaces.
+type ListerDeleter interface {
+	Lister
+	Deleter
+}
+
+// Durations allow the customization of a test timeout and polling interval.
+// They can be customized for each test and/or the default can be changed by updating the package globals.
+type Durations struct {
+	Timeout      time.Duration
+	PollInterval time.Duration
 }
 
-// SetDefaults defaults the struct fields if necessary.
-func (input *ControlplaneClusterInput) SetDefaults() {
-	if input.CreateTimeout == 0 {
-		input.CreateTimeout = 10 * time.Minute
+// SetDefaults is a convenience for library consumers so they never have to specify values for this type.
+func (d *Durations) SetDefaults() {
+	if d.Timeout == 0 {
+		d.Timeout = DefaultTimeout
 	}
+	if d.PollInterval == 0 {
+		d.PollInterval = DefaultPollInterval
+	}
+}
 
-	if input.DeleteTimeout == 0 {
-		input.DeleteTimeout = 5 * time.Minute
+// CreateRelatedResourcesInput is the input type for CreateRelatedResources.
+type CreateRelatedResourcesInput struct {
+	Creator          Creator
+	RelatedResources []runtime.Object
+	Durations
+}
+
+// CreateRelatedResources is used to create runtime.Objects.
```

Comment:
It looks like currently we are limiting these helpers to only being able to be used by tests running with ginkgo, if we also provided a wrapper similar to Gomega's `WithT`, we could allow for these helpers to be called from any type of test.

https://github.com/onsi/gomega/blob/6be6c439588487cd908a3a700795660c2a16dfec/gomega_dsl.go#L348

Reply:
Any reason not to call this just `CreateResources`, it seems like it's generic enough that it doesn't actually need to be resources related to a Control Plane here.

Reply:
First comment: Supporting other test frameworks is outside the scope of the framework, first paragraph here, but it is open to interpretation as it's not very explicit: https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20191016-e2e-test-framework.md#implementation-detailsnotesconstraints

If the framework would benefit from adding support for other test types then we should definitely talk about it. I suspect this would be useful for CAPA? If you can open an issue then we can discuss adding support there. Either way that's out of scope for this PR.

Second comment: This was copy pasta from the old existing code. It was added for CAPV and I didn't want to remove the behavior. It might be that people use this to create everything instead of the specific Create functions, but until we get more signal/usage I'm inclined to leave it as is.

Reply:
Created https://github.com/kubernetes-sigs/cluster-api/issues/2306 to track adding support for using the test framework without ginkgo: https://github.com/kubernetes-sigs/cluster-api/issues/2306